### PR TITLE
[Parser] Make Lexer::getStateForBeginningOfToken() to work with Trivia

### PR DESCRIPTION
**Pull request by rintaro**
CC: @nkcsgexi 

Resolve a problem where `BacktrackingScope` does not restore `Trivia` state.

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: include/swift/Parse/Lexer.h [hunk 1]
```diff
@@ -282,17 +280,16 @@ class Lexer {
...
 
   /// \brief Restore the lexer state to a given one, that can be located either
   /// before or after the current position.
   void restoreState(State S, bool enableDiagnostics = false) {
     assert(S.isValid());
     CurPtr = getBufferPtrForSourceLoc(S.Loc);
-    LeadingTrivia = S.LeadingTrivia;
-    TrailingTrivia = S.TrailingTrivia;
+    LeadingTrivia.clear();
+    TrailingTrivia.clear();
```
Comment (nkcsgexi) [inline include/swift/Parse/Lexer.h]: Are these two lines necessary at all? `LexImpl` will set them, right?

**Review by nkcsgexi — COMMENTED** [review]
Thank you for fixing this! a minor question inline.

File: include/swift/Parse/Lexer.h (hunk 1 above)
Comment (rintaro) [reply include/swift/Parse/Lexer.h]: It's necessary, `lexImpl()` only *append* to them.
Probably, we should put this to the top of `lexImpl()` i.e:

```swift
void Lever::lexImpl() {
  LeadingTrivia.clear()
  TrailingTrivia.clear()
  ...
```
We could do that in another PR?

Comment (nkcsgexi) [reply include/swift/Parse/Lexer.h]: We certainly could. Feel free to merge:)

**Pull request merged**